sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java

```diff
@@ -449,54 +520,155 @@ static RunQueryRequest makeRequest(Query query, @Nullable String namespace) {
 
     @Override
     public void validate(PBegin input) {
-      checkNotNull(getProjectId(), "projectId");
-      checkNotNull(getQuery(), "query");
+      checkNotNull(getProjectId(), "projectId ValueProvider");
+
+      if (getProjectId().isAccessible()) {
+        checkNotNull(getProjectId().get(), "projectId");
+      }
+
+      if (getQuery() == null && getGqlQuery() == null) {
+        throw new IllegalArgumentException(
+            "Either query or gql query ValueProvider should be provided");
+      }
+
+      if (getQuery() != null && getGqlQuery() != null) {
+        throw new IllegalArgumentException(
+            "Only one of query or gql query ValueProvider should be provided");
```

Comment:
I guess it's not too bad, especially since it's an implementation detail - but let's discuss separately whether ValueProvider's should support this use case. ValueProvider's come from PipelineOptions, so having a ValueProvider for the credentials is not at all unreasonable. Can you start a thread on dev@ ?